File: cs/algorithms_in/bucket_sort.cs
namespace algorithms_in;

public class BucketSort
{
    public static void bucketSort(int[] list) {
        int n = int.MinValue;
        for (int i = 0; i < list.Length; i++) {
            if (list[i] > n) {
                n = list[i];
            }
        }

        List<int>[] buckets = new List<int>[n + 1];

        for (int i = 0; i < list.Length; i++) {
            int key = list[i];
            if (buckets[key] == null) {
                buckets[key] = new List<int>();
            }

            buckets[key].Add(list[i]);
        }

        int k = 0;
        for (int i = 0; i < buckets.Count(); i++) {
            if (buckets[i] != null) {
                List<int> bucket = buckets[i];
                for (int j = 0; j < bucket.Count(); j++) {
                    list[k++] = bucket[j];
                }
            }
        }
    }

    public static void bucketSort2(int[] list) {
        int min = list[0];
        int max = list[0];
        for (int i = 1; i < list.Length; i++) {
            if (list[i] < min) {
                min = list[i];
            }
            if (list[i] > max) {
                max = list[i];
            }
        }

        int range = max - min + 1;
        int offset = 0 - min;

        int[] buckets = new int[range];

        for (int i = 0; i < list.Length; i++) {
            int key = list[i] + offset;
            buckets[key] = buckets[key] + 1;
        }

        int k = 0;
        for (int i = 0; i < buckets.Length; i++) {
            if (buckets[i] > 0) {
                for (int j = 0; j < buckets[i]; j++) {
                    list[k++] = i - offset;
                }
            }
        }
    }
}

File: rust/algorithms_in/src/bucket_sort.rs
pub fn bucket_sort(array: &mut [i32]) {
    let mut min: i32 = array[0];
    let mut max: i32 = array[0];

    for i in 1..array.len() {
        if array[i] < min {
            min = array[i];
        }
        if array[i] > max {
            max = array[i];
        }
    }

    let range: usize = (max - min + 1) as usize;
    let offset: i32 = 0 - min;

    let mut buckets: Vec<usize> = vec![0; range];

    for i in 0..array.len() {
        let key: usize = (array[i] + offset) as usize;
        buckets[key] = buckets[key] + 1;
    }

    let mut k: usize = 0;
    for i in 0..buckets.len() {
        if buckets[i] > 0 {
            for _ in 0..buckets[i] {
                array[k] = i as i32 - offset;
                k += 1;
            }
        }
    }
}